### ISSUE TITLE
Waiting on duplicate triggers in a list

### ISSUE THREAD

**Issue by ktbarrett** [cocotb/cocotb#1263]
> Would be interesting to see what happens in that case, but obviously out of scope here.

_Originally posted by @eric-wieser in https://github.com/cocotb/cocotb/pull/1262_

This is interesting...

```python
@cocotb.test(expect_error=cocotb.result.SimTimeoutError)
def test_duplicate_triggers_in_list(dut):

    @cocotb.coroutine
    def writeClkEdge():
        dut.clk.value = 0
        yield cocotb.triggers.Timer(1, 'ns')
        dut.clk.value = 1

    cocotb.fork(writeClkEdge())
    yield [RisingEdge(dut.clk), RisingEdge(dut.clk)]
    # the rising edge event should be handled by this point and because only one edge is written once,
    # the next statement should timeout
    yield cocotb.triggers.with_timeout(RisingEdge(dut.clk), 2, 'ns')
```

This fails as expected.

```python
@cocotb.test(expect_error=cocotb.result.SimTimeoutError)
def test_duplicate_triggers_in_list(dut):
    trigger = RisingEdge(dut.clk)

    @cocotb.coroutine
    def writeClkEdge():
        dut.clk.value = 0
        yield cocotb.triggers.Timer(1, 'ns')
        dut.clk.value = 1

    cocotb.fork(writeClkEdge())
    yield [trigger, trigger]
    # the rising edge event should be handled by this point and because only one edge is written once,
    # the next statement should timeout
    yield cocotb.triggers.with_timeout(trigger, 2, 'ns')
```

This passes...

The `RisingEdge` trigger in the second example is being registered twice and triggers twice. We likely need a way to detect duplicate triggers in a list. But I also think this would cause issues if the duplicates were being yielded in multiple forked coroutines. Will test that later.

**Comment by ktbarrett** [comment]
Nvm, this was an unrelated issue with an incorrect implementation of `with_timeout`.
